### ISSUE TITLE
Break tests/common into a separate crate?

### ISSUE THREAD

**Issue by Apanatshka** [uutils/coreutils#959]
I found your integration test utilities for command-line tools through [this reply](https://users.rust-lang.org/t/whats-everyone-working-on-this-week-32-2016/6824/6) on "What's everyone working on". I'm also [writing tests](https://github.com/BurntSushi/cargo-benchcmp/pull/10) for a command-line tool and your utility functions look very useful! Can you break it out into a separate crate and publish it, so other Rust developers creating command-line tools can benefit? :)

(Or perhaps you can combine work with an existing crate like [assert_cli](https://github.com/killercup/assert_cli), though that would take longer I suppose)

**Comment by wimh** [comment]
For clarification, do you mean [util.rs](https://github.com/uutils/coreutils/blob/master/tests/common/util.rs) and it's dependencies?

Regarding assert_cli, this project is MIT licensed. assert_cli is dual MIT and Apache licensed. That means it is not possible to combine it into assert_cli, unless all authors approve. (The main reason is Apache including a patent grant which MIT does not).

But I also see that cargo-benchcmp is dual licensed MIT and Unlicense. That means you cannot use a MIT-only licensed crate in your project. In particular because the MIT license requires a copyright notice in binary distributions. Unlicense allows binary distributions but does not require a copyright notice. So cargo-benchcmp is not compatible with MIT-only crates.

So basically you cannot use the coreutils utilities for your project.

**Comment by nathanross** [comment]
> So basically you cannot use the coreutils utilities for your project.

I'm glad you are trying to help @Apanatshka but I believe you are providing incorrect information here.

Cargo-benchcmp being dual-licensed with unlicense does not preclude it from using MIT dependencies as long as it is distributed with a proper copyright notice for those dependencies, and in a way that communicates that it is the source code of benchcmp itself, and not its dependencies, that is redistributable under unlicense.

@Apanatshka I'll spin out an (MIT-licensed, obviously) repo of util.rs sometime this week. I can't offer information about whether it will ever be maintained under the auspices of uutils project as that is a much different conversation of whether uutils maintaining something like that adds any value to their project.

**Comment by Apanatshka** [comment]
> I'll spin out an (MIT-licensed, obviously) repo of util.rs sometime this week. 

Great! Thanks 👍 

> I can't offer information about whether it will ever be maintained under the auspices of uutils project as that is a much different conversation of whether uutils maintaining something like that adds any value to the project.

Understood. I think it would make sense for uutils to maintain it since you guys wrote it and are the primary (currently only) users. But ultimately it matters more that it *is* maintained than *who* maintains it.

**Comment by wimh** [comment]
> Cargo-benchcmp being dual-licensed with unlicense does not preclude it from using MIT dependencies as long as any distribution that includes the dependencies includes a proper copyright notice for those dependencies

That is correct, but the Unlicense allows you to distribute binaries **without copyright notice** of the dependencies. If the Unlicense applied only to the source code, there would be no problem. But the Unlicense explicitly allows binary distributions.

Of course it is not a problem until someone distributes binaries (using the Unlicense). And it is also not a problem as long as noone whose copyright is voilated complains. It just that you "promise" something - to the person who is distributing the binaries - which is not true. 

But this is not really my business, as I am not a copyright owner of any code involved. But I just want to make this clear. I created cargo-benchcmp#11 to explain, and leave it to to this.

**Comment by nathanross** [comment]
@wimh I appreciate your passion for the topic.

I think I can see your drift here (that people statically linking binaries based off of both projects may, through an unfamiliarity with licensing, unintentionally violate licenses of dependencies when distributing those binaries). 

As you say, this is a concern more for downstream projects, and less for the spin-out library discussed here,  that must be weighed among others by downstream projects

**Comment by Apanatshka** [comment]
@nathanross: What's the status on this?

**Comment by nathanross** [comment]
Okay, I've got a usable crate up and running. It can be added with ```second_law = "0.2"``` in the Cargo.toml dev-dependencies

[repo link](https://github.com/nathanross/second_law) - I still need to edit down the readme a lot.

[example usage with stainless](https://github.com/nathanross/second-law-test/blob/master/tests/test_stainless.rs)

[example usage with vanilla tests](https://github.com/nathanross/second-law-test/blob/master/tests/test_basic.rs)

If at some point those demo links are broken, a demo will be referenced in the repo link.

Most of the new functionality is around helping the test find the binary, and the fixtures, and be aware of the existence of subcommands. No real documentation yet, but a quick cheatsheet. All references to ```scene.``` functions below are &self functions.

* ```TestScenario``` is now ```Scene```, use ```new_scene!()``` if your binary has the same name as your crate, otherwise provide the binary with ```Scene::new(bin_subpath_relative_to_out_dir:&Path)``` (and by out dir, "target/debug", "target/release", etc.)

* The fixtures directory is divided into two groups, a "root directory" that defaults to "tests/fixtures" and can only be set otherwise once (using ```scene.fixtures_root(:&Path)```), and a "fixtures subdir" PathBuf that can be iteratively appended to (through ```scene.fixtures_subdir : &Path```) for nested test suites.

* The ```scene.subcmd_arg(:&OsString)``` and ```scene.subcmd_args(:&[OsString])``` functions can be used to provide any args added to it to the beginning of each ucmd() call.

This is advantageous as it allows avoiding having to repeat the subcommand every time, without even having to create a separate boilerplate function/closure everytime you want to have subcommands in a test suite.

* The ```scene.multicall(bin_subpath: &Path, subcommand_name: &OsString)``` function is a convenience function meant to provide first-class support for multicall binaries. It desugars to fixtures_subdir(subcommand_name), and subcommand_args(subcommand_name).

It should be called before other subcommand_args or fixtures_subdir calls usually. Although this may make it seem that this means it should be combined with the constructor, its left on its own to allow for multiple depth subcommands.

----

You are getting the bleeding edge version of this. Especially for the configuration functions above, there are probably a few bugs. UCommand remains largely untouched.

The main UX changes of this crate from uutils.rs are as follows: 

- TestScenario remained Scene

- The Fixtures AtPath is now retrieved from ```scene.working_dir()``` instead of scene.fixtures

- The addition of the above configuration functions needed for making the code generalizable to different repos. It took me a while to settle on a design that I felt was generalizable enough yet and ergonomic in a way that any UX breaking changes would be minor (function renames, etc)

**Comment by nathanross** [comment]
Forgot to ping @Apanatshka

**Comment by Apanatshka** [comment]
Thanks a million @nathanross! :smiley: 

I would strongly advise you to start using this new crate in uutils now. But that's up to you. For me, this issue is resolved. I will hopefully have time to try out `second_law` next weekend.
